src/librustc_typeck/check/op.rs

```diff
@@ -393,9 +408,17 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
             Err(()) => {
                 let actual = self.resolve_type_vars_if_possible(&operand_ty);
                 if !actual.references_error() {
-                    struct_span_err!(self.tcx.sess, ex.span, E0600,
+                    let mut err = struct_span_err!(self.tcx.sess, ex.span, E0600,
                                      "cannot apply unary operator `{}` to type `{}`",
-                                     op.as_str(), actual).emit();
+                                     op.as_str(), actual);
+                    let missing_trait = match op {
+                        hir::UnNeg => "std::ops::Neg",
+                        hir::UnNot => "std::ops::Not",
+                        hir::UnDeref => "std::ops::UnDerf"
+                    };
+                    err.note(&format!("an implementation of `{}` might be missing for `{}`",
+                                             missing_trait, operand_ty));
```

Comment:
It would be interesting to check wether `actual` is an unsigned numeric value, `!`, `()` or `str` to avoid showing this note, maybe even emitting a note explaining that unsigned numeric values cannot be negative when `missing_trait == hir::UnNeg`.